# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/ImageEncoder.cpp
# Commit: cd4ca1691c0b
# Full Hash: cd4ca1691c0b74f665e4e43692d04a992d2a11db
# Author: Olli Pettay <Olli.Pettay@helsinki.fi>
# Date: 2022-06-23 03:54:50
# Regressor Bug: 1769878
# File Overlap Count: 1
# Description:
#   Bug 1769878, ensure objects owned by the worker thread are cleared when the worker is shutting down, r=asuth,aosmond
#   
#   Depends on D147858
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D149693
# ==============================================================================

diff -r 851d0144fd91 -r cd4ca1691c0b dom/base/ImageEncoder.cpp
--- a/dom/base/ImageEncoder.cpp	Wed Jun 22 14:06:30 2022 +0000
+++ b/dom/base/ImageEncoder.cpp	Wed Jun 22 14:06:31 2022 +0000
@@ -125,6 +125,11 @@
     return mCreationEventTarget;
   }
 
+  bool CanBeDeletedOnAnyThread() {
+    return !mEncodeCompleteCallback ||
+           mEncodeCompleteCallback->CanBeDeletedOnAnyThread();
+  }
+
  private:
   uint64_t mImgSize;
   nsAutoString mType;
@@ -192,8 +197,10 @@
     rv = mEncodingCompleteEvent->GetCreationThreadEventTarget()->Dispatch(
         mEncodingCompleteEvent, nsIThread::DISPATCH_NORMAL);
     if (NS_FAILED(rv)) {
-      // Better to leak than to crash.
-      Unused << mEncodingCompleteEvent.forget();
+      if (!mEncodingCompleteEvent->CanBeDeletedOnAnyThread()) {
+        // Better to leak than to crash.
+        Unused << mEncodingCompleteEvent.forget();
+      }
       return rv;
     }
 